## Cleaning

In [66]:
import pandas as pd
df = pd.read_csv("name_data/company_person_name_dataset.csv")
df

name  class lang
0                The Canal of the Angels      0   en
1                      Rescue Renovation      0   en
2       Agatha Christie: The ABC Murders      0   en
3                            Siti Akbari      0   ar
4                                  Stany      0   pl
...                                  ...    ...  ...
199995                   Robber's Bridge      0   en
199996                       Johan Renck      0   en
199997                      Lyle Stewart      1   en
199998           Thomas Colclough Watson      1   en
199999                              Gavà      0   ca

[200000 rows x 3 columns]

In [67]:
df = df[df["class"]==1]
#look into smote for imbalance

In [68]:
import numpy as np

In [69]:
print(np.any(df["name"].isnull()))
print(np.any(df["class"].isnull()))
print(np.any(df["lang"].isnull()))

True
False
False


In [70]:
print(len(df.isnull()))

48190


In [71]:
df = df.dropna()

In [72]:
df.drop_duplicates()

name  class lang
11                 Raha Etemadi      1   et
12       Leena Peltonen-Palotie      1   fi
16                  Luma Grothe      1   en
20                Takuya Kakine      1   ja
21               Ōuyáng Zhènhuá      1   mi
...                         ...    ...  ...
199976          Clyde Donaldson      1   en
199980          Terry Alexander      1   en
199989             Neil Roebuck      1   en
199997             Lyle Stewart      1   en
199998  Thomas Colclough Watson      1   en

[47759 rows x 3 columns]

In [73]:
language_counts = df["lang"].value_counts
print("Language distribution:")
print(language_counts)

Language distribution:
<bound method IndexOpsMixin.value_counts of 11        et
12        fi
16        en
20        ja
21        mi
          ..
199976    en
199980    en
199989    en
199997    en
199998    en
Name: lang, Length: 48188, dtype: object>


In [74]:
print(np.any(df["name"].isnull()))
print(np.any(df["class"].isnull()))
print(np.any(df["lang"].isnull()))
print(np.any(df.duplicated()))

False
False
False
True


In [75]:
df = df.drop_duplicates()

In [76]:
print(np.any(df.duplicated()))

False


In [77]:
print("Language distribution:")
print(language_counts)

Language distribution:
<bound method IndexOpsMixin.value_counts of 11        et
12        fi
16        en
20        ja
21        mi
          ..
199976    en
199980    en
199989    en
199997    en
199998    en
Name: lang, Length: 48188, dtype: object>


In [78]:
df["lang"] = df["lang"].str.strip()
language_counts = df["lang"].value_counts()
print("Language distribution:")
print(language_counts)

Language distribution:
lang
en    23352
es     2532
it     1224
fr     1181
ja     1155
      ...  
my        1
sd        1
kk        1
pa        1
si        1
Name: count, Length: 103, dtype: int64


## Additional Cleaning to Names with Special Characters that arent apostrophes, dashes, spaces, or accents.


Reset df index starting from 0

In [79]:
df.reset_index(drop=True, inplace=True)
print(df)

                          name  class lang
0                 Raha Etemadi      1   et
1       Leena Peltonen-Palotie      1   fi
2                  Luma Grothe      1   en
3                Takuya Kakine      1   ja
4               Ōuyáng Zhènhuá      1   mi
...                        ...    ...  ...
47754          Clyde Donaldson      1   en
47755          Terry Alexander      1   en
47756             Neil Roebuck      1   en
47757             Lyle Stewart      1   en
47758  Thomas Colclough Watson      1   en

[47759 rows x 3 columns]


Create langname to find the alphabet name for each character in a name instead of just the first letter (Anna said its better to have full list) and create column

In [133]:
import unicodedata
from unidecode import unidecode

In [81]:
langname = lambda x:[unicodedata.name(char).split(' ')[0] for char in x]
df['alphabet'] = df['name'].apply(langname)
print(df['alphabet'].head())

0    [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...
1    [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...
2    [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...
3    [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...
4    [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...
Name: alphabet, dtype: object


In [82]:
allalphatype = []
#List of all alphabet types of every character in the dataset
for i in df['name']:
    for x in langname(i):
        allalphatype.append(x)

print(len(set(allalphatype)))

#Count 
total = 0
for i in set(allalphatype):
    count = 0
    total +=1
    for x,y in enumerate(df['name']):
        if i in langname(y):
            count += 1
    print(count, i)
print(total)

55
1 ETHIOPIC
2 MIDDLE
5 COMBINING
1 DAGGER
5 NUMBER
1 MYANMAR
10 SEMICOLON
1 LESS-THAN
2 TAMIL
4 EN
180 QUOTATION
3 ARMENIAN
1 FEMININE
1 GURMUKHI
14 GREEK
4 VERTICAL
301 LEFT
184 CJK
85 DIGIT
2 HIRAGANA
17 DEVANAGARI
15 SOLIDUS
4 KANNADA
1 DOLLAR
1 FULLWIDTH
5 BENGALI
47392 LATIN
2 GEORGIAN
1 TELUGU
5 MODIFIER
2 THAI
23 HANGUL
1 SINHALA
13 HEBREW
339 APOSTROPHE
1400 HYPHEN-MINUS
46500 SPACE
1 ZERO
1 KATAKANA
2 IDEOGRAPHIC
1 WHITE
312 RIGHT
1 EQUALS
3 ASTERISK
75 CYRILLIC
1 EXCLAMATION
1 ACUTE
5 AMPERSAND
1 SYRIAC
1 LOW
1 MALAYALAM
77 ARABIC
418 COMMA
2489 FULL
31 COLON
55


In [83]:
def findname(character, indexes):
    for x, y in enumerate(df['name']):
        if str(character) in (langname(y)):
            indexes.append(x)
            print(x, y, df.iloc[x, 2])

def replacechar(orig, repl, indexes):
    for y in indexes:
        df.iloc[y, 0]=df.iloc[y, 0].replace(orig, repl)
        print(df.iloc[y, 0])

Alphabets: (HEBREW, TELUGU, SINHALA, CJK, HIRAGANA, ARMENIAN, DEVANAGARI, KATAKANA, CYRILLIC, KANNADA,BENGALI, GURMUKHI, LATIN, TAMIL, GREEK, ARABIC, GEORGIAN, MALAYALAM, SYRIAC, HANGUL, MYANMAR, THAI, ETHIOPIC) (24)


Non alphabets: (ASTERISK, 418 COMMA, 2489 FULL, 1400 HYPHEN-MINUS, 5 COMBINING, 1 FULLWIDTH, 180 QUOTATION, 10 SEMICOLON, 4 VERTICAl, 31 COLON, 46500 SPACE, 15 SOLIDUS, 1 EXCLAMATION, 1 DOLLAR, 1 DAGGER, 1 LESS-THAN, 1 ACUTE, 1 EQUALS, 1 ZERO, 339 APOSTROPHE, 1 LOW, 2 IDEOGRAPHIC, 1 FEMININE, 5 MODIFIER, 312 RIGHT, 5 AMPERSAND, 1 WHITE, 301 LEFT, 2 MIDDLE, 85 DIGIT, 4 EN, 5 NUMBER) (32)

Keeping all alphabets and APOSTROPHE and HYPHEN-MINUS



### COMMA

In [84]:
com = []
findname('COMMA', com)

28 Vicente de Valverde y Alvarez de Toledo, O.P. es
161 St. Jutta of Kulmsee, T.O.S.F. en
195 Sir William Jardine, 7th Baronet en
439 Ric Manrique, Jr. ca
488 Saint Teresa of Calcutta,MC en
591 Y. Pierre Gobin, MD fr
637 Louis VII, Duke of Bavaria en
976 Samuel D. Pagdilao, Jr. en
1013 Blessed Agnes of Jesus, O.P. en
1025 St. Gerard of Csanád, O.S.B. hu
1193 Thomas Wharton, Jr. en
1332 Simon IV, Lord of Lippe en
1421 Walter D. Druen, Jr. da
1475 Saint Rose Venerini, M.P.V. la
1724 Hubert Duck, Deuteronomy Duck and Louis Duck (Quack Pack) en
2102 Sir Vincent Meredith, Bt en
2250 Eckhard I, Count of Scheyern de
2294 Sir Robert Rich, Bt en
2370 Edward Francis Mickolus, Jr. en
2376 Chen-Yuan Lee, en
2379 Charles Louis, Count of Nassau-Saarbrücken en
2429 Saint Jeanne de Lestonnac,O.D.N. fr
2659 Archie J. Old, Jr. en
2722 Sir Archibald Alison, Bt en
2829 Ellen Heber-Katz, PhD en
2965 Denmark Groover, Jr. en
3254 Henry IV, Count of Bar en
3334 Frederick Landis, Jr. en
3342 Mario Santos, Jr. 

### IDEOGRAPHIC

In [85]:
ide = []
findname('IDEOGRAPHIC', ide)

31381 漢人 陽子　(Kando Yōko) zh-CN
43099 飯島澄男　Sumio Iijima zh-CN


### COMBINING

In [86]:
comb = []
findname('COMBINING', comb)

13038 (Его́р  Лигачёв) tg
16669 Мико́ла Миха́йлович Люби́нський uk
20518 (Дми́трий Я́зов) ru
25538 (Никола́й Вознесе́нский) ru
27501 (Валенти́на Матвие́нко) bg


### ASTERISK

In [87]:
ast = []
findname('ASTERISK', ast)

6816 León Ávalos y Vez* es
12044 Biga*Ranx en
25458 *Tomko sk


### FULLWIDTH

In [88]:
fullw = []
findname('FULLWIDTH', fullw)

43794 Thomas Prence （or，Thomas Prince） en


### QUOTATION

In [89]:
quo = []
findname('QUOTATION', quo)

190 Francis "Frank" Forshew sv
199 Evelyn "Champagne" King en
248 Ernesto "Boy" F. Herrera en
440 Graeme "Shirley" Strachan en
557 Augustine "Willie" Dominguez en
775 Brian "Goldbelt" Maxine en
834 Joyce "Fenderella" Irby en
883 Farley "Jackmaster" Funk en
1686 Edgar "Ted" Codd en
1780 Lewis Burwell "Chesty" Puller cy
2551 Ivan Maček "Matija" hr
3020 Billy "Harp" Hamilton en
3094 George "Chappie" Johnson Jr. en
3285 "Professor" Irwin Corey en
3391 Alfredo "El Salsero" Escalera es
3477 Peter Ulysses "Sturgis" Turner en
3716 Mark Edmund "Duke" Bainum en
4987 Robert W. "Bob" Straub en
4997 James "J.T." Taylor en
5503 Arthur "Andy" B. VanGundy Jr. te
5743 David L. "Dave" Pearce en
6793 William "Bill" L. Mack en
6942 John Richard "Jack" Williams en
7024 Ronald Dean "Ron" Givens en
7480 H. K. "Kenneth" Cassidy en
7725 "Şeker" Ahmed Pasha tr
8116 H. L. "Matty" Matthews en
8296 Abigail "Abby" Sciuto it
8436 Hoffman Lee "Hop" Fuller en
8608 Christopher "Chris" A. Brown en
9339 Donald Howard "Do

### SEMICOLON

In [90]:
semi = []
findname('SEMICOLON', semi)


12404 Caesar Lucius Aurelius Commodus Antoninus Augustus (180); la
15848 Shinmen Takezō; Miyamoto Bennosuke; Niten Dōraku; Shinmen Musashi no Kami Fujiwara no Genshin ja
28415 Lele; Ezequiel haw
33844 Michael O;Shea en
35732 Steven K. D&#39;Arcy en
38925 Darna; Princess Vara; Olga sv
43280 Given name: Xuanyuan (Hsuan-yuan; 軒轅) zh-CN
44626 Marcus Aurelius Numerius Numerianus (from birth to elevation to Caesar); la
45209 Flavius Honorius (from birth to accession); en
46661 (198 to 211); en


### VERTICAL

In [91]:
ver = []
findname('VERTICAL', ver)

2816 thumb|right|Spitfire Mk IIa P7350 of the BBMF is the only existing airworthy Spitfire that fought in the Battle of Britain. en
4319 frameless|center en
14690 Veronika Sramatythumb|Veronika Sramaty making of The Top Ten en
22302 Timmy O'Dowdthumb|Timmy Dowd in Dowdies December 2011 en


### COLON 

In [92]:
col = []
findname('COLON', col)

762 Ancestral name: Jiang (姜) gu
3596 German:  Josef Franz Leopold en
4092 German: Henriette Maria Norberta en
5305 Ardeshir (or in higher Persian: Ardashir) Kamkar en
6336 Lily Munster/ OTS name: Phoebe Munster en
7433 File:Hollidaycamo.jpg en
15053 Farid Aslani (Persian : فرید اصلانی) fa
18358 Family name: Hóu (侯) en
19400 Dutch: Gabriël Boudewijn Karel Maria nl
20732 Dutch: Albert Leopold Clemens Maria Meinrad de
21742 Sajid AliUrdu:ساجد  علی fa
22198 Given name: Unknown en
24154 (Arabic:عبد العزيز بديع صالح السليمي) ar
25809 Chinese: Duàn Sīpíng (段思平) zh-CN
25879 Given name: Dao 導 ja
27432 Given name: Zhoupu (州蒲) en
27498 Makhdoom Khusro Bakhtyar (Seraiki:مخدوم  کسر  بختٻد) fa
27503 Laqab: Qutb ad-Din (shortly) uz
27582 Spanish: María Luisa de Borbón y Sajonia es
28808 Maganti Murali Mohan ( Born : Maganti Raja Babu ) te
31311 Traditional Chinese:韓世忠 zh-CN
34236 Given name: Dawud en
37179 Nu:Tone ro
40523 German: Wilhelm Friedrich Karl Ernst Joachim Albrecht en
40992 Family name:Si

### SOLIDUS

In [93]:
sol = []
findname('SOLIDUS', sol)

6336 Lily Munster/ OTS name: Phoebe Munster en
7610 Paula Baracho / Paula Ribeiro pt
8389 Rajendra Choudhary / राजेन्द्रा चौधरी hi
10940 Amanda Porter/Emily Thorne en
12453 Anne-Marie Baiynd/Presnell/Buttrick/Band en
15445 Líu Hóngdù (劉弘度) / Líu Bīn (劉玢) zh-CN
21785 Elavarasan a/l Elangowan ms
22295 Gay Stanhope Falcon/Gay Laurence/Tom Laurence/Michael "Mike" Waring en
23549 Carsten Bo Eriksen / MBD73 da
29724 Marcy Levy/Marcella Detroit en
29991 Mikhail Shervashidze/Chachba en
31777 Philip Hammond/Hamman, Sr. hi
43349 Nanthakumar a/l Kaliappan ta
45738 Maruja Isabella Sevilla (Spanish Era), Nina Concepcion/Cristy (Reincarnation) es
46497 Victor a/l Andrag ro


### EXCLAMATION

In [94]:
exl = []
findname('EXCLAMATION', exl)

27704 Dr. Kucho! hi


### DOLLAR

In [95]:
dol = []
findname('DOLLAR', dol)

35273 Ty Dolla $ign fi


### DAGGER

In [96]:
dag = []
findname('DAGGER', dag)

7182 Philipp zu Hohenfelden † de


### LESS-THAN

In [97]:
lessth = []
findname('LESS-THAN', lessth)

44921 <span en


### ACUTE

In [98]:
acu = []
findname('ACUTE', acu)

31033 Désiré M´Bonabucya fr


In [99]:
langname('M´Bonabucya')

['LATIN',
 'ACUTE',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN']

### EQUALS

In [100]:
eq = []
findname('EQUALS', eq)

24351 =Giovanni Agostino Marliani it


### ZERO (ZERO WIDTH SPACE)

In [101]:
zer= []
findname('ZERO', zer)

13387 صادق قطب‌زاده fa


### KEEP APOSTROPHE 

In [102]:
apo = []
findname('APOSTROPHE', apo)

147 Lawrence O'Brien en
263 William G. T'Vault en
553 John O'Donoghue en
600 Brendan O'Brien en
686 Breon O'Casey en
882 Detta O'Cathain gd
928 Junior Sa'u hr
1192 Ak'ak'i Meipariani ht
1248 Gary De'Roux fr
1419 Patrick O'Sullivan fi
1548 Sameh El-Saharty'' su
1949 Harriet Osborne O'Hagan en
2140 Cinders O'Brien en
2195 George 'Jocka' Todd en
2461 Rod O'Connor en
2570 Luke O'Connor en
2699 Mononc' Serge fr
2761 (Dato') it
2871 Aodh Ó'Raghallaigh ga
2972 Abdullah Sallum al-Samarra'i ar
2977 Abdillahi Suldaan Mohammed 'Timacade' so
3097 Graba' ro
3230 Steve O'Neill en
3339 Eleni Koiosi' fi
3839 Brittany O'Connell en
4006 Willem 's Gravesande en
4125 Paul O'Grady en
4184 Mark O'Brien en
4208 Anne d'Alençon fr
4297 Harvey O'Brien en
4342 Summer O'Brien en
4343 Martin Patrick O'Connell en
4367 F'Murr en
4405 Molly O'Day en
4420 Bryant O'Dare Hammett, Jr. en
4724 Ann O'Connell en
4746 सुर्नेद्र झा 'सुमन ' hi
4751 Conor Cruise O'Brien en
4782 Joseph Leonard O'Brien en
4887 John O'Neill en
492

### LOW

In [103]:
low = []
findname('LOW', low)

8271 Panpayak_Jitmuangnon jw


### FEMININE

In [104]:
fem = []
findname('FEMININE', fem)

31035 José Mª. Cabral Bermúdez es


### MODIFIER

In [105]:
mod = []
findname('MODIFIER', mod)

19200 al-Ḥasan ibn ʻAlī al-Barbahārī ar
35526 Mulielealiʻi haw
35885 Israel "Iz" Kamakawiwoʻole haw
37882 Wikolia Kamehamalu Keawenui Kaahumanu-a-Kekūanaōʻa, Victoria Kamāmalu Kaahumanu IV haw
41238 Abū al-Ḥassan ʿAlī ibn ʿUthmān al-Marīni ar


In [106]:
langname('ʻAlī al-Barbahārī')

['MODIFIER',
 'LATIN',
 'LATIN',
 'LATIN',
 'SPACE',
 'LATIN',
 'LATIN',
 'HYPHEN-MINUS',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN']

In [107]:
langname('ʿAlī ibn ʿUthmān al-Marīni')

['MODIFIER',
 'LATIN',
 'LATIN',
 'LATIN',
 'SPACE',
 'LATIN',
 'LATIN',
 'LATIN',
 'SPACE',
 'MODIFIER',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'SPACE',
 'LATIN',
 'LATIN',
 'HYPHEN-MINUS',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN',
 'LATIN']

### RIGHT (312) and LEFT (301)

In [108]:
langname('’')


['RIGHT']

In [109]:
rig = []
lef = []
findname('RIGHT', rig)
findname('LEFT', lef)

46 David John (Dave) Shannon en
101 (Armen Lubin) de
197 C.W. (Bill) Hutchins en
355 Koshinaka Makoto (越中睦士) ja
555 Josip Seissel (Jo Klek) sl
762 Ancestral name: Jiang (姜) gu
802 Frank J. O’Connell en
915 (최영일) ko
1357 (Henry Roger) Justin Lewis en
1724 Hubert Duck, Deuteronomy Duck and Louis Duck (Quack Pack) en
1757 Charles D. “Chuck” McAtee en
1794 (GOSEComCComCS) en
1997 Junsei Terasawa (寺沢潤世) ja
2335 () en
2346 Pavel (Pavlo) Gintov sk
2411 (Τίνα Μπιρμπίλη) el
2446 Blessed John Finch (martyr) en
2479 David Hirst (arachnologist) en
2502 Hormisdas Etienne Djibri (Estefan Jabri) fr
2505 Baldwin (II) Rátót hu
2605 Kim Tae-Jin (김태진) ko
2608 (Serdar-ı Ekrem Ömer Paşa) tr
2761 (Dato') it
2942 Claire Kirby (Claire Debrois) en
2980 Saint Honoratus (Honorius) of Amiens la
3187 Milo Gunderson (body) en
3654 Alexander (Oleksandr) Khvostenko-Khvostov ru
4084 Kanysh Satpayev (Kanysh Satbaev) ru
4216 Mohammad Boroujerdi (محمد بروجردی) ar
4301 Örüg Temür Khan (Gulichi) tr
4308 Samuel Robles (Pana

### AMPERSAND

In [110]:
amp = []
findname('AMPERSAND', amp)

4309 Isaac & Miria en
22717 Tom & Olly en
24908 Saints Hor, Besoy, & Daydara bn
35732 Steven K. D&#39;Arcy en
38598 Lanfranchi & Farina it


### WHITE

In [111]:
whi = []
findname('WHITE', whi)

19465 Beauty☆Takaco en


### MIDDLE

In [112]:
mid = []
findname('MIDDLE', mid)

20020 (額爾德特·文繡) zh-CN
47690 (愛新覺羅·胤禟) zh-CN


### DIGIT

In [113]:
dig = []
findname('DIGIT', dig)

195 Sir William Jardine, 7th Baronet en
1024 Auriane Mallo in 2013 la
2519 Coi4 Si1 Bui4 en
2766 Lok6 Dai3 ar
2816 thumb|right|Spitfire Mk IIa P7350 of the BBMF is the only existing airworthy Spitfire that fought in the Battle of Britain. en
2940 Ckay1 vi
3625 26th Clan Chief en
4112 Sir Edward Malet, 4th Bt en
4500 D-Ray 3000 en
4740 Robert de Holland, 1st Baron Holand es
6796 James Younger, 5th Viscount Younger of Leckie en
7745 Kyoko Inoue#5 ja
9952 Mateusz Rutkowski20px pl
12404 Caesar Lucius Aurelius Commodus Antoninus Augustus (180); la
12412 Can4 Jin4-hei1 en
12480 The Chain Gang of 1974 en
12817 William Pakenham, 4th Earl of Longford en
13902 George Douglas, 1st Earl of Dumbarton en
14063 9ice ar
14222 21 Savage en
14966 György Berencsi 3rd hu
15526 J57 ar
15874 1316-Sv.4 en
15914 Sir Charles Hastings, 1st Baronet en
16286 Gu2 Geoi6-gei1 bg
16428 Dik6 Lung4 en
17228 A1 ig
17400 Sir George White, 1st Baronet en
17700 wai6 laan4 ar
17701 1978 en
17727 Alexander Hood, 1st Viscount

In [114]:
for i, x in enumerate(dig):
    if langname(df['name'][x][0]) == ['DIGIT']:
        print(x, df['name'][x], df['lang'][x])

3625 26th Clan Chief en
14063 9ice ar
14222 21 Savage en
15874 1316-Sv.4 en
17701 1978 en
26310 1317-P.80 en
29759 1314-P.39 en
32216 2Mex en
33174 12th Planet en
33732 1315 (1899)P.-2 en
35044 1312-Sv.6 en
39681 2Tap ar
46347 1321-P.44 en
47230 1318-Ağ. Top1 az


### EN

In [115]:
enh = []
findname('EN', enh)

14408 Stevan Nedić–Ćela hr
18503 Jerzy Waldorff–Preyss en
19030 Meinhard II of Gorizia–Tyrol en
38917 James Stirling (1800–1876) en


### NUMBER

In [116]:
num = []
findname('NUMBER', num)

7745 Kyoko Inoue#5 ja
23111 Alacrana Plata #2 es
35599 Masked Burning #2 en
35732 Steven K. D&#39;Arcy en
38372 James Gunnar Nixon #1 en


In [117]:
alphatype1 = []
for i in df['name']:
    for x in langname(i):
        alphatype1.append(x)

print(len(set(alphatype1)))

total1 = 0
for i in set(alphatype1):
    count = 0
    total1 +=1
    for x,y in enumerate(df['name']):
        if i in langname(y):
            count += 1
    print(count, i)
print(total1)

55
1 ETHIOPIC
2 MIDDLE
5 COMBINING
1 DAGGER
5 NUMBER
1 MYANMAR
10 SEMICOLON
1 LESS-THAN
2 TAMIL
4 EN
180 QUOTATION
3 ARMENIAN
1 FEMININE
1 GURMUKHI
14 GREEK
4 VERTICAL
301 LEFT
184 CJK
85 DIGIT
2 HIRAGANA
17 DEVANAGARI
15 SOLIDUS
4 KANNADA
1 DOLLAR
1 FULLWIDTH
5 BENGALI
47392 LATIN
2 GEORGIAN
1 TELUGU
5 MODIFIER
2 THAI
23 HANGUL
1 SINHALA
13 HEBREW
339 APOSTROPHE
1400 HYPHEN-MINUS
46500 SPACE
1 ZERO
1 KATAKANA
2 IDEOGRAPHIC
1 WHITE
312 RIGHT
1 EQUALS
3 ASTERISK
75 CYRILLIC
1 EXCLAMATION
1 ACUTE
5 AMPERSAND
1 SYRIAC
1 LOW
1 MALAYALAM
77 ARABIC
418 COMMA
2489 FULL
31 COLON
55


### Time to Drop

In [118]:
drop = []
def dropfn(list):
    for i in list:
        drop.append(i)

listofdrops= [com, ide, comb, ast, fullw, quo, semi, ver, col, sol, exl, dol, dag, lessth, acu, eq, zer, low, fem, mod, rig, lef, amp, whi, mid, dig, enh, num]
for y in listofdrops:
    dropfn(y)
print(len(drop))
print(len(set(drop)))


1398
996


In [119]:
df = df.drop(drop)

In [120]:
len(df)

46763

In [121]:
df.reset_index(drop=True, inplace=True)
df

name  class lang  \
0                 Raha Etemadi      1   et   
1       Leena Peltonen-Palotie      1   fi   
2                  Luma Grothe      1   en   
3                Takuya Kakine      1   ja   
4               Ōuyáng Zhènhuá      1   mi   
...                        ...    ...  ...   
46758          Clyde Donaldson      1   en   
46759          Terry Alexander      1   en   
46760             Neil Roebuck      1   en   
46761             Lyle Stewart      1   en   
46762  Thomas Colclough Watson      1   en   

                                                alphabet  
0      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
1      [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
2      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
3      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  
4      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  
...                                                  ...  
46758  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
46759  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
46760  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
46761  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
46762  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  

[46763 rows x 4 columns]

### Doing basic cleaning again

In [122]:
print(np.any(df["name"].isna()))
print(np.any(df["name"] == ''))
print(np.any(df["name"] == ' '))
print(np.any(df["name"] == '  '))


False
False
False
False


In [123]:
df['alphabet'] = df['name'].apply(langname)

In [124]:
df.reset_index(drop=True, inplace=True)
df

name  class lang  \
0                 Raha Etemadi      1   et   
1       Leena Peltonen-Palotie      1   fi   
2                  Luma Grothe      1   en   
3                Takuya Kakine      1   ja   
4               Ōuyáng Zhènhuá      1   mi   
...                        ...    ...  ...   
46758          Clyde Donaldson      1   en   
46759          Terry Alexander      1   en   
46760             Neil Roebuck      1   en   
46761             Lyle Stewart      1   en   
46762  Thomas Colclough Watson      1   en   

                                                alphabet  
0      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
1      [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
2      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
3      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  
4      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  
...                                                  ...  
46758  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
46759  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...  
46760  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
46761  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...  
46762  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...  

[46763 rows x 4 columns]

In [125]:
allalphatype1 = []
#List of all alphabet types of every character in the dataset
for i in df['name']:
    for x in langname(i):
        allalphatype1.append(x)

print(len(set(allalphatype1)))

#Count 
total1 = 0
for i in set(allalphatype1):
    count = 0
    total1 +=1
    for x,y in enumerate(df['name']):
        if i in langname(y):
            count += 1
    print(count, i)
print(total1)

26
1 ETHIOPIC
1 MYANMAR
2 TAMIL
2 ARMENIAN
1 GURMUKHI
1 SINHALA
11 GREEK
3 KANNADA
2 HIRAGANA
150 CJK
15 DEVANAGARI
5 BENGALI
46468 LATIN
1 TELUGU
1 THAI
3 HANGUL
10 HEBREW
332 APOSTROPHE
1319 HYPHEN-MINUS
45568 SPACE
1 KATAKANA
45 CYRILLIC
1 SYRIAC
1 MALAYALAM
59 ARABIC
2210 FULL
26


## Feature Creation

### alphabet

completed above^

### avg_token_length (complete)

In [126]:
def average_token_length(name):
    total_length = 0
    total_tokens = 0
    tokens = name.split()

    # Calculate the total length of tokens in the name
    token_length = sum(len(token) for token in tokens)

    # Total length (characters in each word) and total number of tokens (words in name)
    total_length += token_length
    total_tokens += len(tokens)
    # Calculate the average token length
    if total_tokens > 0:
        average_length = total_length / total_tokens
        return average_length
    else:
        return 0 
    
avg_token_length = []
for i in df['name']:
    avg_token_length.append(average_token_length(i))

In [127]:
df['avg_token_length'] = avg_token_length

### num_tokens (complete)

In [128]:
def token_len(name):
    total_tokens = 0
    tokens = name.split()
    # Total number of tokens (words in name)
    total_tokens += len(tokens)
    return total_tokens

In [129]:
token_length = []
for i in df['name']:
    token_length.append(token_len(i))
df['num_tokens'] = token_length

### transliteration

In [145]:
!pip install pykakasi
from pykakasi import kakasi

  Using cached pykakasi-2.2.1-py3-none-any.whl (2.4 MB)
  Using cached jaconv-0.3.4-py3-none-any.whl
  Obtaining dependency information for deprecated from https://files.pythonhosted.org/packages/20/8d/778b7d51b981a96554f29136cd59ca7880bf58094338085bcf2a979a0e6a/Deprecated-1.2.14-py2.py3-none-any.whl.metadata
  Using cached Deprecated-1.2.14-py2.py3-none-any.whl.metadata (5.4 kB)
Using cached Deprecated-1.2.14-py2.py3-none-any.whl (9.6 kB)


In [ ]:
def remove_all_diacritics(text):
  normalized_text = unicodedata.normalize('NFKD', text)
  cleaned_text = ''.join([char for char in normalized_text if not unicodedata.combining(char)])
  return cleaned_text
def transliteration(name):
  kks = kakasi()
  kks.setMode("H", "a")
  kks.setMode("K", "a")
  kks.setMode("J", "a")
  kks.setMode("s", True) # Add space, default: no separator
  kks.setMode("C", False) # Do not capitalize
  kks.setMode("E", "K") # Remove accents
  kks.setMode("r", "Hepburn")
  conv = kks.getConverter()
  # Remove all diacritics before transliteration
  name_without_diacritics = remove_all_diacritics(name)
  return conv.do(name_without_diacritics)
df2['transliteration'] = df2['fullname_lc'].apply(transliteration)

In [134]:
transliteration = []
for i in df['name']:
    transliteration.append(unidecode(i))
df['transliteration'] = transliteration

In [161]:
for i, x in enumerate(df['alphabet']):
    if 'SYRIAC' in x:
        print(df.iloc[i])

name                                       ܡܪܝ ܘܪܓܝܤ ܦܐܝܐܦܝܠܝ ܦܠܟܦܝܠܝ
class                                                               1
lang                                                               en
alphabet            [SYRIAC, SYRIAC, SYRIAC, SPACE, SYRIAC, SYRIAC...
avg_token_length                                                 5.75
num_tokens                                                          4
transliteration                            mry wrgys p'y'pyly plkpyly
Name: 37051, dtype: object


In [165]:
for i, x in enumerate(df['alphabet']):
    if 'ETHIOPIC' in x:
        print(df.iloc[i])

name                                                      ዕምሩ፡ኃይለ፡ሥላሴ
class                                                               1
lang                                                               am
alphabet            [ETHIOPIC, ETHIOPIC, ETHIOPIC, ETHIOPIC, ETHIO...
avg_token_length                                                 11.0
num_tokens                                                          1
transliteration                              `emeru xaayela szelaasee
Name: 25868, dtype: object


In [148]:
jacount = 0
for i, x in enumerate(df['lang']):
    if 'ja' in x:
        jacount+=1
        print(df.iloc[i])
print(jacount)

name                                                    Takuya Kakine
class                                                               1
lang                                                               ja
alphabet            [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...
avg_token_length                                                  6.0
num_tokens                                                          2
transliteration                                         Takuya Kakine
Name: 3, dtype: object
name                                                 Shigeto Kawahara
class                                                               1
lang                                                               ja
alphabet            [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, LAT...
avg_token_length                                                  7.5
num_tokens                                                          2
transliteration                                      Shigeto Kawaha

In [ ]:
df

name  class lang  \
0                 Raha Etemadi      1   et   
1       Leena Peltonen-Palotie      1   fi   
2                  Luma Grothe      1   en   
3                Takuya Kakine      1   ja   
4               Ōuyáng Zhènhuá      1   mi   
...                        ...    ...  ...   
46758          Clyde Donaldson      1   en   
46759          Terry Alexander      1   en   
46760             Neil Roebuck      1   en   
46761             Lyle Stewart      1   en   
46762  Thomas Colclough Watson      1   en   

                                                alphabet  avg_token_length  \
0      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
1      [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...              10.5   
2      [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.0   
3      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.0   
4      [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.5   
...                                                  ...               ...   
46758  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...               7.0   
46759  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...               7.0   
46760  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
46761  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
46762  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               7.0   

       num_tokens          transliteration  
0               2             Raha Etemadi  
1               2   Leena Peltonen-Palotie  
2               2              Luma Grothe  
3               2            Takuya Kakine  
4               2           Ouyang Zhenhua  
...           ...                      ...  
46758           2          Clyde Donaldson  
46759           2          Terry Alexander  
46760           2             Neil Roebuck  
46761           2             Lyle Stewart  
46762           3  Thomas Colclough Watson  

[46763 rows x 7 columns]

### period_freq, dash_freq, apostrophe_freq, space_freq

In [ ]:
df['period_freq'] = df['name'].apply(lambda name: name.count('.'))
df['dash_freq'] = df['name'].apply(lambda name: name.count('-'))
df['apostrophe_freq'] = df['name'].apply(lambda name: name.count('\''))
df['space_freq'] = df['name'].apply(lambda name: name.count(' '))

In [ ]:
df.head()

name  class lang  \
0            Raha Etemadi      1   et   
1  Leena Peltonen-Palotie      1   fi   
2             Luma Grothe      1   en   
3           Takuya Kakine      1   ja   
4          Ōuyáng Zhènhuá      1   mi   

                                            alphabet  avg_token_length  \
0  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.5   
1  [LATIN, LATIN, LATIN, LATIN, LATIN, SPACE, LAT...              10.5   
2  [LATIN, LATIN, LATIN, LATIN, SPACE, LATIN, LAT...               5.0   
3  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.0   
4  [LATIN, LATIN, LATIN, LATIN, LATIN, LATIN, SPA...               6.5   

   num_tokens         transliteration  period_freq  dash_freq  \
0           2            Raha Etemadi            0          0   
1           2  Leena Peltonen-Palotie            0          1   
2           2             Luma Grothe            0          0   
3           2           Takuya Kakine            0          0   
4           2          Ouyang Zhenhua            0          0   

   apostrophe_freq  space_freq  
0                0           1  
1                0           1  
2                0           1  
3                0           1  
4                0           1

## Practice Training 

In [ ]:
#can replace file name later
filename = os.path.join(os.getcwd(), "Name_Of_Origin_Project-", "company.csv")
df = pd.read_csv(filename, header=0)

In [ ]:
y = df['language']
X = df.drop(columns = 'language', axis=1)
X.head()

10% of data as test

In [ ]:
#can change test data size
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=1234)